src/VM/VM.cpp

```diff
@@ -598,6 +598,96 @@ namespace Ark
+                        {
+                            Value* next = popAndResolveAsPtr();
+                            if (next->valueType() != ValueType::List)
+                                throw TypeError("Arguments of concat! should be Lists, got " + types_to_str[static_cast<unsigned>(next->valueType())]);
+
+                            for (auto it = next->list().begin(), end = next->list().end(); it != end; ++it)
+                                list->push_back(*it);
+                        }
+
+                        push(Nil);
+
+                        COZ_PROGRESS_NAMED("ark vm concat!");
+                        break;
+                    }
+
+                    case Instruction::POP_LIST:
+                    {
+                        Value list = *popAndResolveAsPtr();
+                        Value number = *popAndResolveAsPtr();
+
+                        if (list.valueType() != ValueType::List)
+                            throw TypeError("Argument 1 of pop should be a List, got " + types_to_str[static_cast<unsigned>(list.valueType())]);
+                        if (number.valueType() != ValueType::Number)
+                            throw TypeError("Argument 2 of pop should be a Number, got " + types_to_str[static_cast<unsigned>(number.valueType())]);
+
+                        long idx = static_cast<long>(number.number());
+                        idx = (idx < 0 ? list.list().size() + idx : idx);
+                        if (idx >= list.list().size())
+                            throw std::runtime_error("pop: index out of range");
+
+                        list.list().erase(list.list().begin() + idx);
+                        push(list);
+                        break;
+                    }
+
+                    case Instruction::POP_LIST_IN_PLACE:
+                    {
+                        Value *list = popAndResolveAsPtr();
+                        Value number = *popAndResolveAsPtr();
+
+                        if (list->isConst())
+                            throwVMError("can not modify a constant list using `pop!'");
+                        if (list->valueType() != ValueType::List)
+                            throw TypeError("Argument 1 of pop! should be a List, got " + types_to_str[static_cast<unsigned>(list->valueType())]);
+                        if (number.valueType() != ValueType::Number)
+                            throw TypeError("Argument 2 of pop! should be a Number, got " + types_to_str[static_cast<unsigned>(number.valueType())]);
+
+                        long idx = static_cast<long>(number.number());
+                        idx = (idx < 0 ? list->list().size() + idx : idx);
+                        if (idx >= list->list().size())
+                            throw std::runtime_error("pop!: index out of range");
+
+                        list->list().erase(list->list().begin() + idx);
+                        break;
+                    }
+
```

Comment:
The problem behind the refactor of the switch/case of the VM is related to performances, currently this is the fastest method we can use (CTRL+F usually gets you where you need to go if you have the instruction's name)

Reply:
Yeah, these are only my "knee-jerk" reactions because coding practices at work are different. 
I fully agree with going pro-performance and am willing to learn this paradigm.